### PR TITLE
Fix a bug in how "rush build --from" calculates dependents.

### PR DESCRIPTION
**Pull request by qz2017**

### CODE REVIEW

**Comment by pgonzal** [comment]
>   private _dependentList: Map<string, Set<string>>;

Please clarify this data structure... the name is really strange.

```
// For a given project name, this data structure stores a set of other
// project names that are its dependencies, i.e. must be built first.
// The project names are case-sensitive strings including the NPM scope.
private _dependencySetsForProjects
```


---
Refers to: apps/rush/src/actions/RebuildAction.ts:40 in 3db232b. [](commit_id = 3db232bd026e72a9b8eb27711afdcde0fdeaaea1, deletion_comment = False)

File: apps/rush/src/actions/RebuildAction.ts [hunk 1]
```diff
@@ -239,7 +239,7 @@ export default class RebuildAction extends BaseRushAction {
    */
   private _collectAllDependents(project: string): Set<string> {
     const deps: Set<string> = new Set<string>();
-    this._dependentList.get(project)!.forEach((dep) => {
+    (this._dependentList.get(project) || new Set<string>()).forEach((dep) => {
```
Comment (pgonzal) [inline apps/rush/src/actions/RebuildAction.ts]: >new Set<string>() [](start = 41, length = 17)

This is a little clunky.  Compare with:

```
const dependencySet: Set<string> = this.dependencySetsForProjects.get(project);

if (dependencySet) {
  for (const dependency of dependencySet) {
     . . .
  }
}
```

**Comment by pgonzal** [comment]
>       this._rushLinkJson.localLinks[project].forEach(dep => {

Is "dep" supposed to be a "dependency" or a "dependent" (which apparently have opposite meanings in Nick's terminology)?  Our coding standards require full English words everywhere, which might clear this up.

---
Refers to: apps/rush/src/actions/RebuildAction.ts:257 in 3db232b. [](commit_id = 3db232bd026e72a9b8eb27711afdcde0fdeaaea1, deletion_comment = False)

**Comment by pgonzal** [comment]
>   private _dependentList: Map<string, Set<string>>;

OMG I am actually losing my mind trying to understand Nick's terminology here ("dependent" vs "dependency").  I would expect the definitions to be like this:

- If A is a dependency for B (i.e. A needs be built first), then:
- A is a "dependentProject" or just "dependent"
- B is a "dependencyProject" or just "dependency"

This is backed up by the English dictionary:
- dependent: a person who depends on or needs someone or something

---
In reply to: [336221542](https://github.com/Microsoft/web-build-tools/pull/392#issuecomment-336221542) [](ancestors = 336221542)

---
Refers to: apps/rush/src/actions/RebuildAction.ts:40 in 3db232b. [](commit_id = 3db232bd026e72a9b8eb27711afdcde0fdeaaea1, deletion_comment = False)

**Comment by pgonzal** [comment]
>   private _dependentList: Map<string, Set<string>>;

Also please fix up all the places where it says "dep" because that is super ambiguous.  Our coding standards require full English words

---
In reply to: [336225259](https://github.com/Microsoft/web-build-tools/pull/392#issuecomment-336225259) [](ancestors = 336225259,336221542)

---
Refers to: apps/rush/src/actions/RebuildAction.ts:40 in 3db232b. [](commit_id = 3db232bd026e72a9b8eb27711afdcde0fdeaaea1, deletion_comment = False)

**Comment by qz2017** [comment]
>   private _dependentList: Map<string, Set<string>>;

This is dependent set list, meaning the set of things that need to be build after the key package.

---
In reply to: [336225259](https://github.com/Microsoft/web-build-tools/pull/392#issuecomment-336225259) [](ancestors = 336225259,336221542)

---
Refers to: apps/rush/src/actions/RebuildAction.ts:40 in 3db232b. [](commit_id = 3db232bd026e72a9b8eb27711afdcde0fdeaaea1, deletion_comment = False)

**Pull request merged**